### DISCUSSION TITLE
How to deallocate a returned string from Javascript?

### DISCUSSION THREAD

**Discussion by begilbert-sys** [emscripten-core/emscripten#21903] [Q&A]
I have the following function in C++:
```
extern "C" {

char* solve(char* string) {
    char* result = (char*)malloc(strlen(string) * sizeof(char));
    // fill up result 
    return result;
}
}
```

Which I then call from JS like so:
`Module.ccall("solve", "string", ["string"], ["insert string here"])`

There is of course a memory leak because the returned char array never gets deallocated. But I'm unsure of how to deallocate that memory from JS.

**Comment by sbc100** [comment]
I don't believe that is possible with `ccall` since `ccall` automatically converts the returned pointer to a JS string, so you don't have access to the underlying pointer to then call `Module._free()` on it.

Perhaps you could restructure your code so that it returns a pointers to fixed location that gets freed on the next call:

```
char* solve(char* string) {
    static char* result = NULL;
    result = (char*)realloc(result, strlen(string) * sizeof(char));
    // fill up result 
    return result;
}
```
